### ISSUE TITLE
GaudiExec enhancements

### ISSUE THREAD

**Issue by rob-c** [ganga-devs/ganga#764]
GaudiExec loses a few features which it would be good to re-implement:
- [x] Post-processing metadata per subjob
- [ ] Pickling payload before submitting several hundred subjobs
- [x] XML lumi data could be added back in

Edit:
- [ ] Inspect the submitted scripts for `DaVinci().TupleFile`

This list may change but these are the biggest todo items for best support I can think of.

**Comment by rmatev** [comment]
Do you think it makes sense to use `GaudiExec` with released applications, too? I guess you would need to add two properties -- `project` and `version`, and make them mutually exclusive with `directory`. You'd then use `lb-run ${project} ${version} ...` instead of `./run`.

The advantage is that  everything becomes uniform, and you can drop all applications like DaVinci, Moore, etc.
I can submit another issue if you like this proposal and think it deserves it.

**Comment by rob-c** [comment]
@rmatev I've written a function which I was hoping to test/advertise which does a `lb-dev` locally to setup the local environment of a released application.
https://github.com/ganga-devs/ganga/blob/develop/python/GangaLHCb/Lib/Applications/GaudiExecUtils.py#L89
If this does what you want great, if it has bugs, less great but can you file an issue so we can fix it 😄 

The older application within Ganga are there for backwards compatibility in loading old jobs from Ganga 6.1.x and prior so we can't remove these although we have de-facto frozen them and have no intention to support any PR against these moving forwards.

**Comment by rob-c** [comment]
I'm extremely reluctant to add any knowledge of the application name or type to `GaudiExec`. This is information which strictly does not belong there and is due to the new ability to build a reliable target for all LHCb apps.

**Comment by rmatev** [comment]
I'm afraid it does not work with 6.2.0:
```python
Ganga In [26]: prepare_cmake_app('DaVinci', 'v41r2p1')
ERROR    Error: global name 'expandfilename' is not defined
```
(looking at the code, there are some other imports missing, too).

Anyway, it seems to me a bit too much to make an lb-dev project when I don't add code, but I think see your worry. Then, how about doing `directory=$LHCBRELEASES/DAVINCI/DAVINCI_v41r2p1`?

I don't know exactly what magic is needed to run on Dirac, since `$LHCBRELEASES` will resolve to `/afs` by default on LXPLUS, but then again I don't know how lb-dev projects prepared on LXPLUS are able to work on the grid.

**Comment by rob-c** [comment]
@rmatev There needs to be a payload which contains the `./run` and the appropriate `.tar.gz` to use the `GaudiExec` application in this way. Yes it may seem overkill and we can possibly consider placing the application in the `/tmp` directory or something similar if it's a disposable application.
We can consider writing a function which prepares an application object which corresponds to a given LHCb-App and version but I don't want to add any extra information about this to the `GaudiExec` object.

i.e.
```
j=Job()
myApp = makeGaudiExecApp('Davinci', 'v40r2')
j.application=myApp
j.name='DaVinci_v40r2_Job'
j.submit()
```

However I don't want to add this information of a released application if we can help it. All of the correctly resolving `\cvmfs` locations and such in the payload is done in the scripts which LHCb generates for Ganga automatically.
@egede Do you have any thoughts on this?

**Comment by rmatev** [comment]
This might work, but still seems like a workaround. One thing it does not solve is the proliferation of directories in `gangadir/shared` -- if I submit many jobs with the same application/version, I'll get many shared directories. (Maybe the shared dirs can be named according to the hash of the contents and not be replicated, but this is another story).

In fact, looking at `./run` I get why everything works on the grid -- it is calling `lb-run` internally, which redoes the project path resolution. Also, running from a released project is extremely simple, in `./run` just replace
```sh
exec lb-run   --user-area $(cd $(dirname $0) && pwd) DaVinciDev/v41r2p1 "$@"
```
with
```sh
exec lb-run DaVinci v41r2p1 "$@"
```
and you don't need to send any additional files apart from options.
Of course, the question of how to pass DaVinci and v41r2p1 to GaudiExec remains.

**Comment by rmatev** [comment]
I think what I'm really trying to say is that IMHO `buildGangaTarget` (`execCmd` and `cleanGangaTargetArea`) does not really belong to `GaudiExec`. This is a means to get `./run` (and the accompanying directory) but is not the only one.

**Comment by egede** [comment]
@rmatev While I see your argument, I am not that keen on supporting it. It would only work with very simple options files, that do not include other option files.

@rmatev The issue with an exploding number of directories in the shared area is hopefully solved (it was a serious bug in Ganga causing this). Now, you should only get one directory for each prepared application. We could consider to implement some way to purge old shared areas (but this will then prevent you from resubmitting a job where you have done such a purge on).

@rob-c Why do you wan tto re-introduce pickling of options. In my mind it is a weakness in terms of that it is an extra step that is nor strictly required - and I am pretty sure that there are examples where pickling on the user node, followed by unpickling on the worker node is not fully reversible.

**Comment by rob-c** [comment]
@egede I'm thinking of adding a step to test if pickling works to see if the job makes sense. I don't intend to send the result to the WN I mainly just want to add an extra check to help users before they submit several hundred jobs which fail due to a typo or equivalent.

**Comment by egede** [comment]
@rob-c Attractive idea, but if user has fiddled with their PYTHONPATH before starting Ganga, then this might give a false sense of security. To get a clean environment for testing, you almost need to start a virtual machine as we are aware of from earlier discussions. And that seems a bit of an overkill.

**Comment by rmatev** [comment]
@egede I didn't get "It would only work with very simple options files, that do not include other option files.". Can you elaborate please?

By the way, another argument against sending pickles to WNs is that the shared dir becomes much larger 100-200 KB, instead of a few KB today.

**Comment by egede** [comment]
@rmatev If you have a file A that includes file B, that includes file C and D, you will have to provide all of them and make sure that your PYTHONPATH on the WN will find them. The old style pickling of options solved that problem, but as I say it brings with it other issues.

>By the way, another argument against sending pickles to WNs is that the shared dir becomes much larger 100-200 KB, instead of a few KB today.

@rmatev I am afraid that the shared dir in general in the new scheme becomes much bigger. For example for a Gauss job with a new decay file, it will pack up the directory will *all* the decays files, the C++ code, shared  libraries and all. To work in this way is a choice coming from the core software side.

**Comment by rmatev** [comment]
@egede I might be missing something, but aren't option files shipped along `cmake-input-sandbox.tgz` and not inside?

**Comment by egede** [comment]
@rmatev If you set the options file in `GaudiExec` to `['a.opts']`, then, indeed `a.opts` will be shipped separately. But if this file includes other files, it depends on that they are packed up in the `cmake-input-sandbox.tgz` (or in the release area). This is different compared to the old applications like `DaVinci` where we pickled the full set of options after making a "dry" run of the application.

**Comment by rmatev** [comment]
@egede I understand that. My point was that in theory you don't need `cmake-input-sandbox.tgz` to run some option files (given as you showed) on a released project. @rob-c's suggestion is to essentially create an empty `lb-dev` project, which in turn depends on a released project, and give that to `GaudiExec`. I was hoping this intermediate step can be avoided.

**Comment by rob-c** [comment]
@rmatev What are the most urgent features that people want for `GaudiExec`?

**Comment by rmatev** [comment]
I started looking into adding summary.xml / metadata support to GaudiExec.

The first part is to make the job produce `summary.xml` and ship it to the local host. This part is trivial and it works.

The second point is the parsing and merging of the xmls. For old applications, the parsing was done in the WN and the output was shipped back. There is a good reason for this -- the parsing relies on being in the LHCb software environment. I *think* we should keep this behaviour. 

The more complicated thing is merging in case we have a job with subjobs. Again, to merge the xmls and produce the merged metadata, you need to be in the LHCb environment. `execCmd()` comes to mind immediately. However, it relies on the existence of the project source directory. Here is the deeper issue that there is no guarantee that the source directory exists. I may have made a temporary lb-dev project just to pick up a released app version with no local additions. Then I delete it and next time I run ganga, `app.directory` points to a non-existing directory.

I have two solutions in mind
1. Run the merging script with `lb-run LHCb <version> ...`, where version can be "latest" or a fixed number.
2. Modify `execCmd()` to run from the tar-ed project. I'm not sure if there is a better way than untar-ing on every call. I've implemented that and is there is no perceivable delay for a project with no modifications. Currently, I've changed the signature to `execCmd(self, cmd, use_source=False)` to allow for the use case from `buildGangaTarget()`.

Even if you decide to go for 1, 2 is a something to consider as missing (or modified!) source directory it is a problem on its own.

**Comment by rmatev** [comment]
The deeper problem I was referring to is the fact that an application keeps a pointer to an external resource **and** relies on it. Having the directory is fine for keeping track of provenance (for the user), but is otherwise IMHO not needed.

**Comment by rob-c** [comment]
@rmatev 
Firstly:
There is _nothing_ stopping a simpler and more generic method being written than the `execCmd` within the `GaudiExec` app.
This `execCmd` method is _always_ intended to be run against the app on disk the user has configured as they may have local modifications and we can _never_ know they don't have local changes against any LHCb code. If the user chooses to make no local modifications then unfortunately they still need the app in a writable directory to make the targets required by the `GaudiExec` app.
The `execCmd` is __not__ intended to be a be-all/end-all way of accessing LHCb applications/environment, else it would have been a static function and not dependent on the specific application folder the user has requested.

As for generating the XML, yes this needs to be done on the WN from within the env which has been setup by the user to correctly pull in any potentially local changes. This needs to be added as a minor extra step in the `GaudiExec` app and is potentially only a small amount of work. Merging can be done using either the same env as the users app (imo best for consistency) or any other env the user wants to use.

**Comment by rmatev** [comment]
@rob-c I am not sure if other users share my needs, but I think metadata + a checker that checks that you ran on all input data is indispensable (compares the number of processed events to the number of events according to the bookkeeping). I have such a checker at https://github.com/rmatev/ganga-tools/blob/master/GangaLHCbExt/LHCbBookkeepingChecker.py and if you like I'll be happy to put it in ganga.

**Comment by rmatev** [comment]
@rob-c  OK. Where do we put the functionality for running from the prepared app (on the localhost)? Should it be in `execCmd()` with a switch, e.g.
```
def execCmd(self, cmd, use_source=True)
# or
def execCmd(self, cmd, use_prepared=False)
```
or in a separate method?

PS. I don't find the merged summary particularly useful. The most important for me is to run checkers on the subjobs and fail them.

**Comment by rob-c** [comment]
@rmatev From what I understand the main missing feature here is the summary XML not being generated.
This involves me adding this code as a task which is optionally run on `GaudiExec` jobs controlled by some Boolean at the application level.

I'm not sure I understand the problems here.
If you've submitted jobs with a given application folder on disk I would have assumed that Ganga could make use of this folder until the jobs have finished running.
If this is not the case my first reaction is to wonder why but we can change this code to make use of the prepared application which is stored within the Ganga folder prepared directories somewhere.

**Comment by rob-c** [comment]
As for adding a checker I'm more than happy to add a checker once we have the XML being produced/returned correctly. Looking at your code you may wish to take a look at the changes coming in 6.3.0 as there are changes in how various methods talk to DIRAC now from within Ganga.

**Comment by egede** [comment]
@rob-c 
> If you've submitted jobs with a given application folder on disk I would have assumed that Ganga could make use of this folder until the jobs have finished running.

In Ganga we have the principle that when the job is first submitted, it is independent of changes carried out in the locally checked out code. This includes both changing the code in the checked out area or indeed completely removing it. So if we need to use the users environment at the end of the master job (for merging the XML), then in principle the only "correct" thing to do is to unpack the tarball from the shared area and run the code in that environment. A very heavy operation but in principle the only correct one,

**Comment by rob-c** [comment]
@egede OK I'll code up some something to make use of the correct environment from the unpacked tarball. This should be relatively trivial but will take a little time to check it's working correctly.
